### PR TITLE
[Frontend] [Core] perf: Automatically detect vLLM-tensorized model, update `tensorizer` to version 2.9.0

### PR DESCRIPTION
**Pull request by sangstar**
# Automatically detect vLLM-tensorized model, update `tensorizer` to version 2.9.0
This PR accomplishes several things:

- Updates docstrings to account for tensorizer refactor in #4097 in the `tensorize_vllm_examples.py` example script, and slight corrections to the docstrings of the new, refactored functions.
- **Allows models to be automatically inferred as a vLLM-tensorized model**. Accomplishes this by placing a meta-tensor "footprint" in the serialized model, and removing it at runtime. `vllm_tensorized` as an arg has been removed.
- Updates `tensorizer` to the full release of 2.9.0. 

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by sangstar** [comment]
@Yard1 @ywang96 

Some QoL improvements for Tensorizer and some corrected docstrings, and an update for tensorizer as version 1.9.0 is officially released. No longer need to specify if a model is vLLM-tensorized beforehand, as I've implemented a way for this to be inferred implicitly by registering a meta tensor into the model during serialization with a `vllm-tensorized-marker` and removing it during deserialization.

**Comment by ywang96** [comment]
Will take a look once I have some bandwidth - thanks for the continuous contribution to vLLM!

File: examples/tensorize_vllm_model.py [hunk 1]
```diff
@@ -71,26 +68,30 @@
...
+              model_loader_extra_config=TensorizerConfig(
+                    tensorizer_uri = path_to_tensors,
+                    num_readers=3,
+                    )
+              )
+            
+A serialized model can be used during model loading for the vLLM OpenAI
+inference server. `model_loader_extra_config` is exposed as the CLI arg
+`--model-loader-extra-config`, and accepts a JSON string literal of the
+TensorizerConfig arguments desired.
```
Comment (bbrowning) [inline examples/tensorize_vllm_model.py:85]: Thanks for updating these docs! It took me a bit to figure out the right CLI args to pass before this PR, and this makes it much clearer for future users!

Comment (sangstar) [reply examples/tensorize_vllm_model.py:85]: My pleasure! `examples/tensorize_vllm_model.py` is meant to serve as the unofficial Tensorizer-for-vLLM usage guide, so I made sure to add it to the Examples Tests in the CI suite to make sure it continues working with refactors. I'm not really sure how best to enforce that the docstrings in this example script explain any refactored changes, but the hope is that that will be carried through when ensuring the script continues to work.

File: vllm/model_executor/model_loader/loader.py [hunk 2]
```diff
@@ -285,16 +285,17 @@ def _get_weights_iterator(
...
         return tensorizer_weights_iterator(tensorizer_args)
 
-    def _load_model_unserialized(
+    def _load_model_serialized_cpu(
             self, model_config: ModelConfig, device_config: DeviceConfig,
             lora_config: Optional[LoRAConfig],
             vision_language_config: Optional[VisionLanguageConfig]
     ) -> nn.Module:
-        """Load an unserialized model with tensorizer.
+        """Load a serialized model with tensorizer to the CPU.
```
Comment (ywang96) [inline vllm/model_executor/model_loader/loader.py:293]: Why only CPU? Don't we have `with torch.device(device_config.device):`?

**Review by ywang96 — COMMENTED** [review]
Thank you @sangstar for the continuous contribution! I left some questions.

File: vllm/model_executor/model_loader/tensorizer.py [hunk 3]
```diff
@@ -366,3 +360,57 @@ def tensorizer_weights_iterator(
...
+    if (".vllm_tensorized_marker" in deserializer):
+        return True
+    return False
+
+
+def get_pretensorized_vllm_model(engine: "LLMEngine") -> nn.Module:
+    model = (engine.model_executor.driver_worker.model_runner.model)
+    model.register_parameter(
+        "vllm_tensorized_marker",
+        nn.Parameter(torch.tensor((1, ), device="meta"), requires_grad=False))
```
[the remark this answers is not in the document]
Comment (sangstar) [reply vllm/model_executor/model_loader/tensorizer.py:391]: That's correct. Users would need to re-serialize their models. I could return the `--vllm-tensorized` parameter and warn that it's now deprecated if desired.

File: vllm/model_executor/model_loader/loader.py (hunk 2 above)
Comment (sangstar) [reply vllm/model_executor/model_loader/loader.py:293]: If the model is not vLLM-tensorized, trying to load the model to any other device destination other than CPU is not currently supported, even if `device_config.device` isn't for CPU. `TensorDeserializer` accepts a `device` parameter, and loads the model directly to that device. If loaded directly to the GPU, vLLM's loading architecture then needs to make the model a "vLLM model" (by instantiating its bespoke model layers for the supported loaded model and copying the weights to those instantiated layers on `device_config.device`). Tensorizer loads directly to the target device before it allows loaded weights to be `yield`ed, so if the target device is to the GPU, passing them off to vLLM's loading architecture causes a double load. I tried to circumvent this in my original PR by doing in-place transfers of weights from loaded to instantiated models rather than `torch.copy_`ing, but this isn't possible with vLLM's weight concatenation layers, which require copying.

I circumvented this by directly allowing for serializing ready-to-serve vLLM models, and loading non-vLLM models (like a HF `PreTrainedModel`) to the CPU to allow for vLLM to instantiate it and load it to the GPU without double-copying.

File: examples/tensorize_vllm_model.py (hunk 1 above)
Comment (sangstar) [reply examples/tensorize_vllm_model.py:85]: Left a comment on that!

**Comment by sangstar** [comment]
@ywang96 Resolved comments! Let me know if anything else is needed.

File: vllm/model_executor/model_loader/tensorizer.py [hunk 4]
```diff
@@ -222,13 +233,6 @@ def add_cli_args(
             help="The endpoint for the S3 bucket. Can also be set via the "
             "S3_ENDPOINT_URL environment variable.",
         )
-        group.add_argument(
-            "--vllm-tensorized",
-            action="store_true",
-            help="If enabled, indicates that the serialized model is a vLLM "
-            "model. This is used to determine the behavior of the "
-            "TensorDeserializer when loading tensors from a "
-            "serialized model.")
```
Comment (ywang96) [inline vllm/model_executor/model_loader/tensorizer.py:231]: Should we bring this back and add a depreciation warning?

Comment (sangstar) [reply vllm/model_executor/model_loader/tensorizer.py:231]: We could, but with the refactor from #4097 tensorizer args can't be exposed from the CLI directly anymore anyway.  They can only be encoded in `--model-loader-extra-config`

File: vllm/model_executor/model_loader/loader.py (hunk 2 above)
Comment (ywang96) [reply vllm/model_executor/model_loader/loader.py:302]: I see - this makes sense, one thing I'm worried about is under certain scenario this will cause unexpected CPU RAM OOM, but I would imagine that as a rare case.

I do think we should add some documentation for `Tensorizer` in our main documentation page around its usage and on a high level how it works.

**Comment by sangstar** [comment]
@ywang96 Resolved comments! CI build seems to be waiting on the AMD Tests agent.

**Review by ywang96 — APPROVED** [review]
🚀 LGTM!

**Comment by sangstar** [comment]
@ywang96 Checks passed and ready to merge! 😄
